### DISCUSSION TITLE
Performance difference between USM pointers and buffers/accessors

### DISCUSSION THREAD

**Discussion by blinkfrog** [AdaptiveCpp/AdaptiveCpp#1520] [Q&A] [answered]
Hello,

I have been using SYCL for a while, using exclusively buffers and accessors model due to its higher-level abstraction, safety features, and the potential for runtime optimization through the dependency graph. However, I've read recommendations from Aksel Alpay (@illuhad) about the advantages of USM pointers, particularly regarding their lower latency compared to buffers and accessors.

From my initial benchmarks, which, however, focused mainly on access time, I haven't noticed any performance difference between using USM pointers and buffers/accessors. However, I have not extensively tested the impact on data transfer times, which might be crucial.

Could you provide more details or specific scenarios where USM pointers might indeed offer performance advantages over the buffer and accessor model? I'd like to conduct some benchmarks and, probably, move to USM model, if I would see a significant difference.

Thank you for your guidance and support!

**Comment by illuhad** [comment, marked as answer]
There definitely are cases where there's no observable difference. But if there is a difference, it is almost always in favor of USM. By USM I mean `sycl::malloc_device` with explicit data transfers using `sycl::memcpy` and potentially `sycl::malloc_host` if further data transfer optimizations are needed.

There are several main reasons for this:
* Runtime latency. I don't mean access latency in kernels, but host-side overheads, i.e. kernel submission latency. You will definitely notice that if you submit many short-running kernels where runtime overheads on the host play a role. This was e.g. benchmarked in this paper and is a well-understood effect: https://dl.acm.org/doi/fullHtml/10.1145/3648115.3648120 The fundamental reason is that USM allows the programmer to hardwire their knowledge about dependencies and necessary data transfers at compile time, whereas with buffers the SYCL implementation has to figure it out at runtime, for every single kernel launch. We also have specific optimizations for latency-bound cases like instant submission mode which however can only be enabled when no buffers are used.
* Secondly, accessors are not light-weight objects and they can contribute to register pressure. This can cause slowdowns for large kernels that are limited by resource usage/register pressure. For example: Every single accessor needs to store the allocation size. If you pass in multiple accessors to different buffers of the same size (which is a common pattern, since often we want to have multiple buffers of  a common problem size), the problem size will be duplicated and stored in each of the accessors. The compiler cannot know at compile time that in such a scenario the allocation sizes will all going to be the same, so they will all need their own registers. In the USM model the programmer can express that all allocations are of the same size by just capturing a single variable for the problem size.
* The USM model exposes control over device-accessible, page-locked host memory ("pinned memory" in CUDA) using `sycl::malloc_host` which can be used to further optimize data transfers. Buffers may or may not use that optimization but in general will have to be more conservative because there may be OS limitations to how much memory can be obtained using `malloc_host`.

The buffer model can do optimizations automatically like running independent kernels concurrently, or overlapping data-transfers and kernels. But again this comes at the cost of additional runtime overheads since the SYCL implementation will have to figure out at runtime how to do this. With USM, the programmer can do similar optimizations manually if needed, and while it may be more work, the result will always outperform buffers. This is because runtime overheads will be lower and the programmer can inject their knowledge about how the application actually behaves and what good execution strategies are, rather than having the SYCL runtime guess things.

Another thing to note is that the buffer-accessor model is extremely complex to implement and maintain. You are much, much more likely to hit performance issues or bugs there compared to USM. Just yesterday I found a case where the scheduler would not insert dependencies correctly if one buffer is used on multiple devices :( I will probably have a fix within the next couple of days.

**Reply by blinkfrog** [reply]
Thank you very much for a detailed explanation.

I think I need to make another benchmarks.

Submitting multiple of short kernels, actually, really causes performance problems in two of my projects, and if using USM pointers would help then I surely need to move to pointers.

It just feels so awkward to use them - like a step backwards. I really like how I can use RAII with buffers, and like their reference-counted nature. I even am tempted to write a wrapper around USM pointer which would do reference-counting and release RAM automatically, and also such things as convenient 2D-access, but I afraid that all this would make this wrapper not as lightweight so I can lose benefits of using raw pointers over buffers and accessors.

**Reply by illuhad** [reply]
No problem :)

> Submitting multiple of short kernels, actually, really causes performance problems in two of my projects, and if using USM pointers would help then I surely need to move to pointers.

In that case I would recommend reading our performance guide on latency-bound problems. There are a couple of additional tricks that you may want to combine with USM to get the best effect:
https://github.com/AdaptiveCpp/AdaptiveCpp/blob/develop/doc/performance.md#strong-scalinglatency-bound-problems

> It just feels so awkward to use them - like a step backwards. I really like how I can use RAII with buffers, and like their reference-counted nature. I even am tempted to write a wrapper around USM pointer which would do reference-counting and release RAM automatically, and also such things as convenient 2D-access, but I afraid that all this would make this wrapper not as lightweight so I can lose benefits of using raw pointers over buffers and accessors.

I can understand that. I think unless you constantly create/destroy allocations, I would not expect a lof of overhead from an RAII wrapper, as this would not impact the submission process. Just make sure that you only copy the contained pointer into your kernel, not the entire RAII wrapper. Otherwise you might get overheads, or even UB if the kernel itself tries to free memory.

You can also use `mdspan` with USM pointer in device code which would give you back your 2D access convenience. If you only construct the `mdspan` in device code, you could still express e.g. that several `mdspan` objects should be constructed using the same problem size kernel argument. So it might still be better than accessors.
This will not work though if you construct `mdspan` on the host, as optimizing across the boundary between host and device code is seriously limited. Our JIT compiler (`--acpp-targets=generic`) can do constant propagation across the host-device boundary to some extent, but I think you are using `--acpp-targets=cuda` which cannot do this.
